csg/share/scripts/inverse/check_csg_xml.py

```diff
@@ -0,0 +1,118 @@
+    description = """
+    Check the csg XML file for invalid tags.
+
+    It works by comparing to the tags in the default XML file.
+    """
+    parser = argparse.ArgumentParser(description=description)
+    parser.add_argument(
+        "csg_xml_file",
+        type=argparse.FileType("r"),
+        help="csg XML file",
+    )
+    parser.add_argument(
+        "csg_xml_defaults_file",
+        type=argparse.FileType("r"),
+        help="csg XML defaults file",
+    )
+    # parse
+    args = parser.parse_args()
+    return args
+
+
+def check_votca_settings_xml(root, root_defaults):
+    def iter_xml(node, path="."):
+        yield node, path
+        for child in node:
+            child_path = f"{path}/{child.tag}"
+            for child, child_path in iter_xml(child, path=child_path):
+                yield child, child_path
+
+    sim_progs = ["gromacs", "lammps", "hoomd", "espresso"]
+    found_bad_paths = []
+    for child, child_path in iter_xml(root):
+        # options for cg.bonded are partially listed in cg.non-bonded
+        child_path_non_bonded = child_path.replace("/bonded/", "/non-bonded/")
+        # options for cg.$program are listed in cg.sim_prog, similar per interaction
+        child_path_sim_prog = child_path
+        for sim_prog in sim_progs:
+            child_path_sim_prog = child_path_sim_prog.replace(
+                f"inverse/{sim_prog}/", "inverse/sim_prog/"
+            )
+        if (
+            root_defaults.find(child_path) is None
+            and root_defaults.find(child_path_non_bonded) is None
+            and root_defaults.find(child_path_sim_prog) is None
+            # per group settings can have any name
+            and not child_path.startswith("./inverse/imc/")
+        ):
+            found_bad_paths.append(child_path)
+
+    # find what the user maybe meant
+    suggestions = []
+    for found_bad_path in found_bad_paths:
+        found_bad_tag = found_bad_path.split("/")[-1]
+        try:
+            suggestion = next(
+                default_path
+                for default_element, default_path in iter_xml(root_defaults)
+                if default_element.tag == found_bad_tag
+            )
+            suggestion = suggestion.replace("/non-bonded", "/{bonded,non-bonded}")
+        except StopIteration:
+            suggestion = None
+        suggestions.append(suggestion)
+
+    # output bad tags and suggestions
+    if len(found_bad_paths) > 0:
+        found_bad_lines = ""
+        for found_bad_tag, suggestion in zip(found_bad_paths, suggestions):
+            found_bad_lines += found_bad_tag
+            if suggestion is not None:
+                found_bad_lines += f"  ( did you mean {suggestion} )"
+            found_bad_lines += "\n"
+        print(
+            f"The settings XML file contains the tags\n\n{found_bad_lines}\nbut "
+            "those paths does not exist in the XML defaults file and are therefore "
+            "not supported."
+        )
```

Comment:
Will it fail or just warn?

Reply:
It will fail. `inverse.sh` fails if the Python script outputs something.

Reply:
Hmm sometimes users overwrite scripts and add their own options, so a warning would be better.